app/models/page/DashboardPage.scala

```diff
@@ -16,111 +16,251 @@
 
-  case object Step3 extends Step {
-    def isReached(p: Progress): Boolean = {
-      val ot = p.onlineTest
-      List(ot.onlineTestAwaitingAllocation, ot.onlineTestAllocationConfirmed, ot.onlineTestAllocationUnconfirmed,
-        p.failedToAttend,
-        p.assessmentScores.entered, p.assessmentScores.accepted).contains(true)
+    case object Step1 extends Step {
+      // Step 1 is always reached
+      def isReached(p: Progress): Boolean = true
     }
-  }
-  case object Step4 extends Step {
-    def isReached(p: Progress): Boolean =
-      List(p.assessmentCentre.awaitingReevaluation, p.assessmentCentre.passed, p.assessmentCentre.failed).contains(true)
-  }
 
-  object Step {
-    def determineStep(progress: Option[Progress]): Step = {
-      val step = progress.map { p =>
-        val latestStep: Step = if (Step4.isReached(p)) {
-          Step4
-        } else if (Step3.isReached(p)) {
-          Step3
-        } else if (Step2.isReached(p)) {
-          Step2
-        } else {
-          Step1
-        }
+    case object Step2 extends Step {
+      def isReached(p: Progress): Boolean = {
+        val ot = p.onlineTest
+        List(p.submitted, ot.onlineTestInvited, ot.onlineTestStarted, ot.onlineTestCompleted, ot.onlineTestExpired,
+          ot.onlineTestFailed, ot.onlineTestFailedNotified).contains(true)
+      }
+    }
 
-        latestStep
+    case object Step3 extends Step {
+      def isReached(p: Progress): Boolean = {
+        val ot = p.onlineTest
+        List(ot.onlineTestAwaitingAllocation, ot.onlineTestAllocationConfirmed, ot.onlineTestAllocationUnconfirmed,
+          p.failedToAttend, p.assessmentScores.entered, p.assessmentScores.accepted).contains(true)
       }
+    }
 
-      step.getOrElse(Step1)
+    case object Step4 extends Step {
+      def isReached(p: Progress): Boolean =
+        List(p.assessmentCentre.awaitingReevaluation, p.assessmentCentre.passed, p.assessmentCentre.failed).contains(true)
     }
-  }
 
-  def fromUser(user: CachedData)(implicit request: RequestHeader, lang: Lang): DashboardPage = status(user) match {
-    case Some(ApplicationStatus.WITHDRAWN) => withdrawn(user)
-    case _ => activeApplication(user)
+    object Step {
+      def determineStep(progress: Option[Progress]): Step = {
+        val step = progress.map { p =>
+          if (Step4.isReached(p)) {
+            Step4
+          } else if (Step3.isReached(p)) {
+            Step3
+          } else if (Step2.isReached(p)) {
+            Step2
+          } else {
+            Step1
+          }
+        }
+        step.getOrElse(Step1)
+      }
+    }
   }
 
-  private def withdrawn(user: CachedData)(implicit request: RequestHeader, lang: Lang) = {
-    val progress = user.application.map(_.progress)
-    val latestStepBeforeWithdrawn = Step.determineStep(progress)
+  private def isApplicationSubmittedAndNotWithdrawn(user: CachedData)(implicit request: RequestHeader, lang: Lang) =
+    WithdrawApplicationRole.isAuthorized(user)
```

Comment:
We use roles to control access to the endpoints and I have found that challenging to mantain when we reuse them other different places like this. I also think some of the checks we do in the Role are redundant (isActive, etc) as we are already in the page and to access the page that has been checked already. In RolesUtils there are a few methods that check the progress.